Added Machine ID CLI and configuration references. (#11096)
gravitational/teleport: docs/pages/machine-id/reference/configuration.mdx

    ca_path: "/path/to/ca.pem"

    # CA Pins used to validate the identity of the Teleport Auth Server on first connect.
    ca_pins:
    - "sha256:1111111111111111111111111111111111111111111111111111111111111111"
    - "sha256:2222222222222222222222222222222222222222222222222222222222222222"

# Storage defines where Machine ID internal data is stored.
storage:
    # Directory to store internal bot data. Access to this directory should be
    # limited.
    directory: /var/lib/teleport/bot
    
    # Alternatively, internal data can be stored in memory. "directory" and
    # "memory" are mutually exclusive.
    memory: true

# Destinations specifies where short-lived certificates are stored.
destinations:
    # Directory specifies where short-lived certificates are stored.
    - directory: /opt/machine-id
        # Configure symlink attack prevention. Requires Linux 5.6+.
        # Possible values:
        #   * try-secure (default): Attempt to securely read and write certificates
        #     without symlinks, but fall back (with a warning) to insecure read
        #     and write if the host doesn't support it.

>         #   * secure: Attempt to securely read and write certificates, hard error

Reviewer: ```suggestion
        #   * secure: Attempt to securely read and write certificates, with a hard error
```

        #     if unsupported.
        #   * insecure: Quietly allow symlinks in paths.
        symlinks: try-secure
        
        # Configure ACL use. Requires Linux with a file system that supports ACLs.
        # Possible values:
        #   * try (default on Linux): Attempt to use ACLs, warn at runtime if ACLs
        #     are configured but invalid.
        #   * off (default on non-Linux): Do not attempt to use ACLs.
        #   * required: Always use ACLs, produce a hard error at runtime if ACLs
        #     are invalid.
        acls: try
        
        # One of more roles to grant to the bot. It must have been granted (at
        # least) these roles with `tctl bots add --roles=...`
        # By default, all possible roles are included.
        

>         # Subset of roles allowed during creation `tctl bots add --roles=...`. Can

Reviewer: ```suggestion
        # Subset of roles allowed during creation via `tctl bots add --roles=...`. Can
```

---
title: Machine ID Configuration Reference
description: Configuration reference for Teleport Machine ID.
---

```yaml
# Debug enables verbose logging to stderr.
debug: true

# Address of the Teleport Auth Server (on-prem installs) or Proxy Server
# (Cloud installs).
auth_server: "auth.example.com:3025" # or "example.teleport.sh:443" for Teleport Cloud

# TTL of short-lived machine certificates.
certificate_ttl: "5m"

# Interval at which short-lived certificates are renewed; must be less than
# the certificate TTL.
renewal_interval: "1m"

# If set, quit after the first renewal.
oneshot: true

> # Onboarding values only used on first connect.

Reviewer: ```suggestion
# Onboarding values are only used on first connect.
```

onboarding:
    # Cluster join method. Can be "token" or "iam".
    join_method: "token"

>     # Token used to join cluster.

Reviewer: ```suggestion
    # Token used to join the cluster.
```

    token: "00000000000000000000000000000000"

>     # CA Pin used to validate the identity of Teleport Auth Server on first connect.

Reviewer: ```suggestion
    # CA Pin used to validate the identity of the Teleport Auth Server on first connect.
```

    ca_path: "/path/to/ca.pem"

>     # CA Pins used to validate the identity of Teleport Auth Server on first connect.

Reviewer: ```suggestion
    # CA Pins used to validate the identity of the Teleport Auth Server on first connect.
```

    ca_pins:
    - "sha256:1111111111111111111111111111111111111111111111111111111111111111"
    - "sha256:2222222222222222222222222222222222222222222222222222222222222222"

# Storage defines where Machine ID internal data is stored.
storage:
  # Directory to store internal bot data. Access to this directory should be
  # limited.
  directory: /var/lib/teleport/bot

# Destinations specify one or more impersonated certificates sets to generate.
destinations:

>   # `directory` is the only destination supported in the initial iteration;

Reviewer: ```suggestion
  # Directory is the only destination supported in the initial iteration;
```
to be consistent with the way other fields are named in this page

---
title: Machine ID Configuration Reference
description: Configuration reference for Teleport Machine ID.
---

```yaml
# Debug enables verbose logging to stderr.
debug: true

# Address of the Teleport Auth Server (on-prem installs) or Proxy Server
# (Cloud installs).
auth_server: "auth.example.com:3025" (On-Prem)

> auth_server: "example.teleport.sh:443" (Cloud)

Reviewer 2: I don't know how we prefer to write mutually-exclusive alternatives in our docs, but the duplicate `auth_server` entries here might cause confusion.

# TTL of short-lived machine certificates.
certificate_ttl: "5m"

# Interval at which short-lived certificates are renewed; must be less than
# the certificate TTL.
renewal_interval: "1m"

# If set, quit after the first renewal.
oneshot: true

# Onboarding values only used on first connect.
onboarding:
    # Cluster join method. Can be "token" or "iam".
    join_method: "token"
    # Token used to join cluster.
    token: "00000000000000000000000000000000"
    # CA Pin used to validate the identity of Teleport Auth Server on first connect.
    ca_pin: "sha256:1111111111111111111111111111111111111111111111111111111111111111"
    # CA Pins used to validate the identity of Teleport Auth Server on first connect.
    ca_pins:
    - "sha256:1111111111111111111111111111111111111111111111111111111111111111"
    - "sha256:2222222222222222222222222222222222222222222222222222222222222222"

# Storage defines where Machine ID internal data is stored.
storage:
  # Directory to store internal bot data. Access to this directory should be
  # limited.

>   directory: /var/lib/teleport/bot

Reviewer 2: If we're aiming to be exhaustive, it might be worth mentioning you can alternatively use in-memory storage too:

```yaml
storage:
  # Directory to store internal bot data. Access to this directory should be
  # limited.
  directory: /var/lib/teleport/bot

  # Alternatively, internal data can be stored in memory. Other storage types
  # (e.g. `directory`) should be commented out.
  # memory: true
```

(memory storage is probably appealing right now for non-IAM, one-shot bots since this will never write the renewable cert to disk)

# Destinations specify one or more impersonated certificates sets to generate.
destinations:
  # `directory` is the only destination supported in the initial iteration;
  # the tbot process must be given permissions to write to this directory.
  - directory: /opt/machine-id
    # One of more roles to grant to the bot. It must have been granted (at
    # least) these roles with `tctl bots add --roles=...`
    # By default, all possible roles are included.
    roles: [a, b, c]
    # Which types of certificates to generate. `ssh` is the default.
    kinds: [ssh]
    # A list of configuration templates to generate and write to the
    # destination. Defaults are decided based on the certificate kinds:
    # `ssh` implies `ssh-client` unless manually (not) configured.
    configs:
      # ssh-client generates known_hosts and an ssh_config that can be
      # included. We can ensure the correct certificate kinds are generated
      # while generating the config templates.

>       - ssh-client

Author: Tim: I pulled this straight from the RFD. Is this still accurate?

Reviewer 2: There's only one small error in the above (will need to update the RFD!): `ssh-client` should be `ssh_client`.  That example is otherwise valid but no longer exhaustive.

`directory` now has a verbose form that lets you configure ACL and symlink enforcement:
```yaml
directory:
  path: /opt/machine-id

  # Configure symlink attack prevention. Requires Linux 5.6+.
  # Possible values:
  #   try-secure (default): attempt to securely read and write certificates without symlinks, but fall back (with a warning) to insecure read and write if the host doesn't support it
  #   secure: attempt to securely read and write certificates, hard error if unsupported
  #   insecure: quietly allow symlinks in paths
  symlinks: try-secure

  # Configure ACL use. Requires Linux with a filesystem that supports ACLs.
  # Possible values:
  #   try (default on Linux): attempt to use ACLs, warn at runtime if ACLs are configured but invalid
  #   off (default on non-Linux): do not attempt to use ACLs
  #   required: always use ACLs, produce a hard error at runtime if ACLs are invalid
  acls: try
```

Both are also valid for `storage.directory` entries but only `symlinks` has an effect there. We don't warn about or enforce ACLs at runtime for the bot's internal storage (at least not yet).

The `ssh_client` template also has a verbose form that allows you to override the Teleport proxy port if it's using a non-standard value:
```yaml
configs:
  # ssh-client generates known_hosts and an ssh_config that can be
  # included. We can ensure the correct certificate kinds are generated
  # while generating the config templates.
  - ssh_client:
      proxy_port: 1234

```

---
title: Machine ID Configuration Reference
description: Configuration reference for Teleport Machine ID.
---

```yaml
# Debug enables verbose logging to stderr.
debug: true

> # Address of the Teleport Auth Server (on-prem installs) or Proxy Server

Reviewer: ```suggestion
# Address of the Teleport Auth Server (on-prem installs) or Teleport Cloud tenant.
```

> # (Cloud installs).

Reviewer: ```suggestion
```

auth_server: "auth.example.com:3025" # or "example.teleport.sh:443" for Teleport Cloud

# TTL of short-lived machine certificates.
certificate_ttl: "5m"

# Interval at which short-lived certificates are renewed; must be less than
# the certificate TTL.
renewal_interval: "1m"

# If set, quit after the first renewal.
oneshot: true

# Onboarding values are only used on first connect.
onboarding:
    # Cluster join method. Can be "token" or "iam".
    join_method: "token"

>     # Token used to join the cluster.

Reviewer 3: ```suggestion
    # Token used to join the cluster. (only required for join_method: token)
```

    token: "00000000000000000000000000000000"
    # CA Path used to validate the identity of the Teleport Auth Server on first connect.
    ca_path: "/path/to/ca.pem"
    # CA Pins used to validate the identity of the Teleport Auth Server on first connect.
    ca_pins:
    - "sha256:1111111111111111111111111111111111111111111111111111111111111111"
    - "sha256:2222222222222222222222222222222222222222222222222222222222222222"

# Storage defines where Machine ID internal data is stored.
storage:
  # Directory to store internal bot data. Access to this directory should be
  # limited.
  directory: /var/lib/teleport/bot

  # Alternatively, internal data can be stored in memory. "directory" and
  # "memory" are mutually exclusive.

>   memory: true

Reviewer 3: Any concern about people copy-pasting this file, changing the tokens/endoints, and not seeing this comment and hitting an error on first start?

Maybe we should comment out `# memory: true` so that the example is still there but the config itself is valid.

Author: @timothyb89 mentioned the same, but everywhere else in our reference documentation we document all the options available without commenting out any blocks, even if they are mutually exclusive.

https://goteleport.com/docs/setup/reference/config/

# Destinations specifies where short-lived certificates are stored.
destinations:
  # Directory specifies where short-lived certificates are stored.
  - directory: /opt/machine-id
    # Configure symlink attack prevention. Requires Linux 5.6+.
    # Possible values:
    #   * try-secure (default): Attempt to securely read and write certificates
    #     without symlinks, but fall back (with a warning) to insecure read

>     #     and write if the host doesn't support it.

Reviewer: ```suggestion
    #     and write if the host doesn't support this.
```
Using a vaguer pronoun so it's clearer that it stands for "Attempt to securely read and write certificates without symlinks"
